### PR TITLE
POR-1902 validate name against existing porter apps

### PR DESCRIPTION
**Pull request by ianedwards**
## What does this PR do?

- Client and server side validation for app names
- Ensures that an app name cannot be used twice in the same project
- Also makes sure that name is valid (lowercase characters, numbers, and "-")

### CODE REVIEW

File: api/server/handlers/porter_app/validate.go [hunk 1]
```diff
@@ -112,11 +112,25 @@ func (c *ValidatePorterAppHandler) ServeHTTP(w http.ResponseWriter, r *http.Requ
...
+		c.HandleAPIError(w, r, apierrors.NewErrPassThroughToClient(err, http.StatusInternalServerError))
+		return
+	}
+
 	if appProto.Name == "" {
 		err := telemetry.Error(ctx, span, nil, "app proto name is empty")
 		c.HandleAPIError(w, r, apierrors.NewErrPassThroughToClient(err, http.StatusBadRequest))
 		return
 	}
+	for _, existingApp := range existingApps {
```
Comment (ferozemohideen) [inline api/server/handlers/porter_app/validate.go:127]: let's log the app name to telemetry before this check

File: dashboard/src/main/home/app-dashboard/create-app/CreateApp.tsx [hunk 2]
```diff
@@ -481,6 +481,8 @@ const CreateApp: React.FC<CreateAppProps> = ({ history }) => {
     }
   }, [porterApps, name.value]);
 
+  console.log("errors", errors);
```
Comment (ferozemohideen) [inline dashboard/src/main/home/app-dashboard/create-app/CreateApp.tsx:484]: remove

File: internal/repository/gorm/porter_app.go [hunk 3]
```diff
@@ -34,6 +34,16 @@ func (repo *PorterAppRepository) ListPorterAppByClusterID(clusterID uint) ([]*mo
 	return apps, nil
 }
 
+func (repo *PorterAppRepository) ListPorterAppsByProjectID(projectID uint) ([]*models.PorterApp, error) {
```
Comment (ferozemohideen) [inline internal/repository/gorm/porter_app.go:37]: does this need to check deployment target id?

**Review by ferozemohideen — APPROVED** [review]

File: internal/repository/gorm/porter_app.go (hunk 3 above)
Comment (ianedwards) [reply internal/repository/gorm/porter_app.go:38]: An app from the same project can be deployed to multiple deployment targets. It just has to be unique across the project

**Pull request merged**